uma/utils/tlv_utils.go

```diff
@@ -0,0 +1,168 @@
+package utils
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+)
+
+// BytesCodable is an interface for types that can be marshaled to and unmarshaled from bytes.
+type BytesCodable interface {
+	MarshalBytes() ([]byte, error)
+	UnmarshalBytes([]byte) error
+}
+
+// TLVCodable is an interface for types that can be marshaled to and unmarshaled from TLV.
+type TLVCodable interface {
+	MarshalTLV() ([]byte, error)
+	UnmarshalTLV([]byte) error
+}
+
+// MarshalTLV marshals a struct to TLV.
+// It will marshals all the field with tag "tlv".
+// The "tlv" tag value will be the type of the field.
+func MarshalTLV(v interface{}) ([]byte, error) {
+	val := reflect.ValueOf(v)
+	if val.Kind() != reflect.Ptr || val.Elem().Kind() != reflect.Struct {
+		return nil, fmt.Errorf("marshal requires a pointer to a struct")
+	}
+
+	val = reflect.Indirect(val)
+	typ := val.Type()
+
+	var result []byte
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		tag := typ.Field(i).Tag.Get("tlv")
+		if tag == "" {
+			continue
```

Comment:
Will this just silently drop those fields? Do you think it should it be an error?

Reply:
One example is that if we ever want to use 2 bytes as type, we can specify it in the struct without TLV field.